File: ICPC/WF2022/c.py
[n, m] = [int(x) for x in input().split(' ')]

limit = 1
powers = [1]*3985
while (powers[limit]//limit) < m:
    limit +=1
    powers[limit] = limit ** n

found = False
for q in range(2, limit+1):
    for p in range(1, q):
        if (m % ((powers[q] - powers[q-p])//p) ) == 0:
            print(p, q)
            found = True
        
        if found:
            break
    if found:
        break

if not found:
    print("impossible")

File: ICPC/WF2022/c.cpp
#include <iostream>
using namespace std;
typedef int64_t ll;

int main(){
    ll n, m; 
    cin>>n>>m;

    int limit = 1;
    bool foundLimit = false;
    ll powers[3985];
    powers[1] = 1;
    int power;
    while(!foundLimit){
        ++limit;
        power = 1;

        powers[limit] = 1;
        while( power < n && m/powers[limit] >= limit){
            powers[limit] *= limit;
            power += 1;
        }
        
        if(power == n && power
    }
    
    ll pResp, qResp;
    bool found = false;
    for(ll q = 2; !found && q <= limit ; ++q){
        for(ll p = 1; !found && p < q; ++p){
            if (m % (powers[q] - powers[q-p]) == 0){
                 pResp = p;
                 qResp = q;
                 found = true;
            }
        }
    }
    if(!found){
        cout<<"impossible\n";
    }
    else{
        cout<<pResp<<" "<<qResp<<"\n";
    }
    return 0;
}